### DISCUSSION TITLE
Trouble with Type Safety in Hono Zod Validation Middleware

### DISCUSSION THREAD

**Discussion by riculum** [honojs/hono#2670] [Q&A]
I've been working on implementing a validation middleware in Hono using Zod and `@hono/zod-validator`. Here's the code snippet of the middleware:

    const validationMiddleware = (schema: ZodSchema, target: keyof ValidationTargets) => {
        return createMiddleware(async (c, next) => {
            await next()
            zValidator(target, schema, (result, c) => {
                if (!result.success) {
                    return c.json({
                        success: false,
                        error: {
                            code: 400,
                            message: result.error.issues[0].message,
                            innerError: {
                                timestamp: new Date(Date.now())
                            }
                        }
                    })
                }
            })
        })
    }

And I use it like this:

    validationMiddleware(userValidation, 'json')

The issue I'm facing is that I can't seem to get type safety on the return value. Specifically, when I try to use `c.req.valid('json')`, TypeScript throws the following error:

    TS2345: Argument of type string is not assignable to parameter of type never on the 'json' input

I'm puzzled about why TypeScript is inferring never instead of the expected type. Could someone please advise on how to resolve this and ensure type safety in this scenario? Thanks!

**Comment by NicoPlyley** [comment]
Are you trying to access it in your route? If so your `await next()` is before the validation meaning that would run after the route returns

**Reply by riculum** [reply]
Yes, inside a POST route, like:
```
    const app = new Hono().post('/', validationMiddleware(userValidation, 'json'), async (c) => {
            const user = c.req.valid('json')
```
So, are you suggesting that I need to adjust the placement of `await next()`? Just to clarify, my issue lies in not receiving the type of the returned validation, specifically in the case of `userValidation`. Consequently, `const user` is inferred as type `never`. When I apply `c.req.valid('json')` to `'json' `directly, I encounter the error: `"TS2345: Argument of type string is not assignable to parameter of type never."`

**Reply by NicoPlyley** [reply]
Based on the code you provided the validation is not running at all. Therefore there have been no types passed to the content.
Here is the order for writing middleware:
```ts
app.use(async (c, next) => {
  // Any code here runs before the route handler
  await next(); // Lets the route handler run
  // Code here runs after there is a return from the route handler
});
```
In your case, your code is never validated because it runs after the router handler is returned.
